### PR TITLE
snowsql: Add support for darwin platforms

### PR DESCRIPTION
**Pull request by andehen**
## Description of changes

Add support for darwin platforms (x86-64 and arm64)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
+{
+  lib,
+  stdenv,
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:3]: Given that this package isn't built, but rather installs from a pre-built binary  use of `stdenvNoCC` is likely preferable.
```suggestion
  stdenvNoCC,
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  version,
+  meta,
+  pname,
+}: let
+  arch =
+    if stdenv.isAarch64
+    then "arm64"
+    else "x86_64";
+in
+  stdenv.mkDerivation rec {
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:20]: There is a preference to use `finalAttrs` rather than `rec`. Note that the newly opened parenthesis needs to be closed at the end of the package (see suggestion).
```suggestion
  stdenvNoCC.mkDerivation (finalAttrs: {
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    '';
+
+    postFixup =
+      if arch == "arm64"
+      then ''
+        install_name_tool -change /usr/lib/libSystem.B.dylib ${darwin.Libsystem}/lib/libSystem.B.dylib $out/bin/snowsql
+        install_name_tool -change /usr/lib/libz.1.dylib ${zlib}/lib/libz.1.dylib $out/bin/snowsql
+      ''
+      else "";
+  }
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:56]: Close parenthesis opened before `finalAttrs` on the `stdenvNoCC.mkDerivation) line above.
```suggestion
  })
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      cp -r zipped-content/* $out/bin/
+    '';
+
+    postFixup =
+      if arch == "arm64"
+      then ''
+        install_name_tool -change /usr/lib/libSystem.B.dylib ${darwin.Libsystem}/lib/libSystem.B.dylib $out/bin/snowsql
+        install_name_tool -change /usr/lib/libz.1.dylib ${zlib}/lib/libz.1.dylib $out/bin/snowsql
+      ''
+      else "";
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:55]:
```suggestion
    postFixup = lib.optionalString stdenvNoCC.isAarch64 ''
      install_name_tool -change /usr/lib/libSystem.B.dylib ${darwin.Libsystem}/lib/libSystem.B.dylib $out/bin/snowsql
      install_name_tool -change /usr/lib/libz.1.dylib ${zlib}/lib/libz.1.dylib $out/bin/snowsql
      '';
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    };
+
+    nativeBuildInputs = [unar cpio unzip gzip zlib];
+
+    unpackPhase = ''
+      unar $src
+    '';
+
+    buildPhase = ''
+      cd *snowsql-${version}-darwin_${arch}
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:39]:
```suggestion
      cd *snowsql-${finalAttrs.version}-darwin_${stdenvNoCC.hostPlatform.darwinArch}
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    if stdenv.isAarch64
+    then "arm64"
+    else "x86_64";
+in
+  stdenv.mkDerivation rec {
+    inherit pname version meta;
+    dontStrip = true;
+
+    src = fetchurl {
+      url = "https://sfc-repo.snowflakecomputing.com/snowsql/bootstrap/${lib.versions.majorMinor version}/darwin_${arch}/snowsql-${version}-darwin_${arch}.pkg";
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:25]:
```suggestion
      url = "https://sfc-repo.snowflakecomputing.com/snowsql/bootstrap/${lib.versions.majorMinor finalAttrs.version}/darwin_${arch}/snowsql-${finalAttrs.version}-darwin_${stdenvNoCC.hostPlatform.darwinArch}.pkg";
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  zlib,
+  version,
+  meta,
+  pname,
+}: let
+  arch =
+    if stdenv.isAarch64
+    then "arm64"
+    else "x86_64";
+in
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:19]: The local variable `arch` can be replaced with `stdenv.hostPlatform.darwinArch` or `stdenv.targetPlatform.darwinArch`:
```suggestion
}:
```

I'd recommend removing the indentation or even better format the files using [nixfmt](https://github.com/NixOS/nixfmt): `nix run github:NixOS/nixfmt pkgs/applications/misc/snowsql/darwin.nix`

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  stdenv.mkDerivation rec {
+    inherit pname version meta;
+    dontStrip = true;
+
+    src = fetchurl {
+      url = "https://sfc-repo.snowflakecomputing.com/snowsql/bootstrap/${lib.versions.majorMinor version}/darwin_${arch}/snowsql-${version}-darwin_${arch}.pkg";
+      sha256 =
+        if arch == "arm64"
+        then "acb2096b87466f0fdbff544d9f64feafc23cbee1bdf963e84b9658511ade9536"
+        else "9727c07fc11b1d8adf4a4eb0b5b996d82cd3d9da191a86f4c7b772726f8e5e92";
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:29]: I think using `stdenvNoCC.isAarch64` is more accurate and future-proof here:
```suggestion
      sha256 =
        if stdenvNoCC.isAarch64
        then "acb2096b87466f0fdbff544d9f64feafc23cbee1bdf963e84b9658511ade9536"
        else "9727c07fc11b1d8adf4a4eb0b5b996d82cd3d9da191a86f4c7b772726f8e5e92";
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    nativeBuildInputs = [unar cpio unzip gzip zlib];
+
+    unpackPhase = ''
+      unar $src
+    '';
+
+    buildPhase = ''
+      cd *snowsql-${version}-darwin_${arch}
+      cat snowsql-darwin.pkg/Payload | gzip -d | cpio -idmv
+      unzip SnowSQL.app/Contents/MacOS/snowsql-${version}-darwin_${arch}.zip -d zipped-content
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:41]:
```suggestion
      unzip SnowSQL.app/Contents/MacOS/snowsql-${finalAttrs.version}-darwin_${stdenvNoCC.hostPlatform.darwinArch}.zip -d zipped-content
```

File: pkgs/applications/misc/snowsql/darwin.nix [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    '';
+
+    installPhase = ''
+      mkdir -p $out/bin
+      cp -r zipped-content/* $out/bin/
+    '';
+
+    postFixup =
+      if arch == "arm64"
```
Comment (afh) [inline pkgs/applications/misc/snowsql/darwin.nix:50]:
```suggestion
      if stdenvNoCC.isAarch64
```

File: pkgs/applications/misc/snowsql/default.nix [hunk 11]
```diff
@@ -1,49 +1,20 @@
...
-, rpmextract
-, patchelf
-, makeWrapper
-, openssl
-, libxcrypt-legacy
-}:
-
-stdenv.mkDerivation rec {
+{
+  stdenv,
```
Comment (afh) [inline pkgs/applications/misc/snowsql/default.nix:2]: Given that this package isn't built, but rather installs from a pre-built binary use of stdenvNoCC is likely preferable.
```suggestion
  stdenvNoCC,
```

File: pkgs/applications/misc/snowsql/default.nix [hunk 12]
```diff
@@ -1,49 +1,20 @@
...
     license = licenses.unfree;
-    maintainers = with maintainers; [ andehen ];
-    platforms = [ "x86_64-linux" ];
+    maintainers = with maintainers; [andehen];
+    platforms = ["x86_64-linux" "x86_64-darwin" "aarch64-darwin"];
     mainProgram = "snowsql";
   };
-}
+in
+  if stdenv.isDarwin
```
Comment (afh) [inline pkgs/applications/misc/snowsql/default.nix:18]:
```suggestion
  if stdenvNoCC.isDarwin
```

File: pkgs/applications/misc/snowsql/linux.nix [hunk 13]
```diff
@@ -0,0 +1,45 @@
+{
+  lib,
+  stdenv,
```
Comment (afh) [inline pkgs/applications/misc/snowsql/linux.nix:3]: Given that this package isn't built, but rather installs from a pre-built binary use of stdenvNoCC is likely preferable.
```suggestion
  stdenvNoCC,
```

File: pkgs/applications/misc/snowsql/linux.nix [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+  rpmextract,
+  makeWrapper,
+  openssl,
+  libxcrypt-legacy,
+  zlib,
+  version,
+  pname,
+  meta,
+}:
+stdenv.mkDerivation rec {
```
Comment (afh) [inline pkgs/applications/misc/snowsql/linux.nix:14]: There is a preference to use finalAttrs rather than rec. Note that the newly opened parenthesis needs to be closed at the end of the package (see suggestion).
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/applications/misc/snowsql/linux.nix [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    patchelf \
+      --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+        $out/lib/snowsql
+  '';
+
+  postFixup = ''
+    makeWrapper $out/lib/snowsql $out/bin/snowsql \
+      --set LD_LIBRARY_PATH "${libPath}":"$out/lib"
+  '';
+}
```
Comment (afh) [inline pkgs/applications/misc/snowsql/linux.nix:45]: Close parenthesis opened before finalAttrs on the `stdenvNoCC.mkDerivation) line above.
```suggestion
})
```

File: pkgs/applications/misc/snowsql/linux.nix [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+  nativeBuildInputs = [rpmextract makeWrapper];
+
+  unpackPhase = ''
+    rpmextract $src
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin/
+    mkdir -p $out/lib/
```
Comment (afh) [inline pkgs/applications/misc/snowsql/linux.nix:29]: `mkdir` accepts multiple directories and creates them accordingly and bash, which is the default shell for the nix phases, offers brace expansion as a shorthand
```suggestion
    mkdir -p $out/{bin,lib}
```

**Comment by afh** [comment]
Result of `nixpkgs-review pr 320328` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>snowsql</li>
  </ul>
</details>

**Review by afh — CHANGES_REQUESTED** [review]
Thanks for this, @andehen, much appreciated 👍 

Please find below a few comments and suggestions for improvement.
Hopefully I've been able to explain well what the intent of the suggestions are and provide helpful context, if not please ask me to clarify 🙂

File: pkgs/applications/misc/snowsql/linux.nix (hunk 13 above)
Comment (andehen) [reply pkgs/applications/misc/snowsql/linux.nix:3]: I tried this, but run into issues with the `preFixup` step with `patchelf`, as it depends on the dynamic linker which I guess is included with `stdenv`. Do you have any recommendations for solving that?

**Comment by andehen** [comment]
> Thanks for this, @andehen, much appreciated 👍
> 
> Please find below a few comments and suggestions for improvement. Hopefully I've been able to explain well what the intent of the suggestions are and provide helpful context, if not please ask me to clarify 🙂

Thanks a lot for your review! Really appreciate that you spent time doing this. Writing nix packages is new to me and I learned a lot 👍

File: pkgs/applications/misc/snowsql/linux.nix (hunk 13 above)
Comment (afh) [reply pkgs/applications/misc/snowsql/linux.nix:3]: I wasn't aware of that, thanks, @andehen. Unfortunately I don't have any recommendations for that, so how about leaving `stdenv` in then?

**Review by afh — COMMENTED** [review]
Thanks for considering and accepting the suggestions, @andehen, much appreciated. And glad to hear you found the comments helpful.

File: pkgs/applications/misc/snowsql/linux.nix (hunk 13 above)
Comment (andehen) [reply pkgs/applications/misc/snowsql/linux.nix:3]: Yeah, that's probably the eaisest then.

**Comment by afh** [comment]
@AndersonTorres would be willing to lend your expertise on this one too? It seems the changes in this PR might be related to what is being talked about in #325562.

**Comment by AndersonTorres** [comment]
Well, this is an unfree, binary-only package, and some of observations I did before for vlc and pcsx2 do not apply.

However, I believe it is not really interesting to "merge" both codes in a same expressions, because they are different. There is few to no benefit on "factoring the common terms"  since the common term is too small.

Further, it falls on the previous problem: the Linux and MacOS binaries follow distinct paths, and the maintainers of one have no interest in maintain the other (in a legal sense, something like "I have the tools to test and approve a PR").

Yes, I have noticed the maintainers are the same in this case, but the general principle still applies.

Suppose that a a third person wants to contribute by sending a PR updating only the Apple package, since theey has no Linux to test it. It would be counterproductive to block such a PR.

**Comment by afh** [comment]
>There is few to no benefit on "factoring the common terms" since the common term is too small.

I find this 👆 very helpful as a general guidance, thanks @AndersonTorres.

Regarding maintainers, would setting the maintainers by platforms be useful? E.g.:

```nix
meta {
  # …
  maintainers = with lib.maintainers; [
    ] ++ lib.optionals stdenv.isDarwin [
      afh
    ] ++ lib.optionals stdenv.isLinux [
      AndersonTorres
    ];
}
```

I'm not sure I completely understand your example of a third person wanting to contribute to the Apple package. Let me rephrase in my own words to see if I did:
If a single package that needs to be written very differently per platform, due to the installation method on the platform, it is possible that an update to the Apple part could break the Linux part, yet the Apple contributors/maintainers may not be able to check the Linux part and the Linux contributors/maintainers may not be able to check the Apple part, creating a situation where there are unnecessary complications to move forward with the proposed changes?
Did I get that right?
